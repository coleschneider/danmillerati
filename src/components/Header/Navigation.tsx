import React from "react";
import styled from "styled-components";
import posed from "react-pose";
import { Link, RouteComponentProps } from "react-router-dom";
import LogoIcon from "./logo";
import devices from "../../theme/devices";

const ListContainer = styled.ul`
    text-align: left;
    list-style: none;

    @media ${devices.desktop} {
        width: 100%;
        max-width: 1200px;
        display: flex;
        padding: 0px 2em;
        margin: 0px auto;
    }

    padding: 6em 0px 0px 1em;
`;
const LogoWrapper = styled.li`
    position: absolute;
    top: 0px;
    left: 1em;
    margin-right: auto;
    @media ${devices.desktopMin} {
        left: 2em;
    }
    @media ${devices.desktop} {
        position: relative;
        top: auto;
        left: auto;
        margin-left: 0px;
    }
`;
const LogoLink = styled(Link)`
    position: relative;
    font-weight: bold;
    text-transform: uppercase;
    font-size: 1.5em;
    color: ${({ theme: { colors } }) => colors.black};
    text-decoration: none;
    transition: opacity 0.3s ease 0s;
    border-bottom: 1px solid transparent;
    @media ${devices.desktop} {
        font-size: 0.85em;
        letter-spacing: 0.8px;
        height: 56px;
        padding: 0px 0px 0.1em;
    }
`;

const Item = posed.li({
    open: { x: 0, opacity: 1 },
    closed: { x: -100, opacity: 0 }
});
const MenuItem = styled(Item)`
    opacity: 0;
    @media ${devices.desktop} {
        display: inline-block;
        transform: none !important;
        opacity: 1 !important;
        transition: none 0s ease 0s;
        margin: 0px 0px 0px 1.25em;
    }
`;

const Logo = ({
    isFixed,
    iconType,
    viewBox
}: Partial<HeaderProps> & { iconType: IconTypes; viewBox: string }) => (
    <LogoWrapper>
        <LogoLink to="/">
            <LogoIcon
                viewBox={viewBox}
                iconType={iconType}
                height="56px"
                isFixed={isFixed}
            />
        </LogoLink>
    </LogoWrapper>
);

export interface HeaderLink {
    text: "Home" | "About" | "Contact" | "Gallery";
    to: string;
}
export const headerLinks: HeaderLink[] = [
    { text: "Home", to: "/" },
    { text: "About", to: "/about" },
    { text: "Contact", to: "/contact" },
    { text: "Gallery", to: "/gallery" }
];

const Navigation = ({
    isOpen,
    isFixed,
    toggle,
    history
}: RouteComponentProps & HeaderProps) => {
    const handleClick = (e: React.SyntheticEvent, to: string) => {
        e.preventDefault();
        toggle();
        history.push(to);
    };

    return (
        <ListContainer>
            <Logo
                viewBox="0 0 194.69 53.44"
                iconType="Logo"
                isFixed={isFixed}
            />
            {headerLinks.map(({ text, to }) => (
                <MenuItem key={to} pose={isOpen ? "open" : "closed"}>
                    <Link
                        onClick={e => handleClick(e, to)}
                        className={isFixed ? "stick" : undefined}
                        to={to}
                    >
                        {text}
                    </Link>
                </MenuItem>
            ))}
        </ListContainer>
    );
};
export default Navigation;
