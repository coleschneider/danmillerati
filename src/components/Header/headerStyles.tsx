import * as React from "react";
import styled, { css } from "styled-components";

import posed from "react-pose";
import { Link } from "react-router-dom";
import devices from "../../theme/devices";

export const ListContainer = styled.ul`
    text-align: left;
    list-style: none;

    @media ${devices.desktop} {
        width: 100%;
        max-width: 1200px;
        display: flex;
        padding: 0px 2em;
        margin: 0px auto;
    }

    padding: 6em 0px 0px 1em;
`;
export const LogoWrapper = styled.li`
    position: absolute;
    top: 0px;
    left: 1em;
    margin-right: auto;
    @media ${devices.desktopMin} {
        left: 2em;
    }
    @media ${devices.desktop} {
        position: relative;
        top: auto;
        left: auto;
        margin-left: 0px;
    }
`;
export const LogoLink = styled(Link)`
    position: relative;
    font-weight: bold;
    text-transform: uppercase;
    font-size: 1.5em;
    color: ${({ theme: { colors } }) => colors.black};
    text-decoration: none;
    transition: opacity 0.3s ease 0s;
    border-bottom: 1px solid transparent;
    @media ${devices.desktop} {
        font-size: 0.85em;
        letter-spacing: 0.8px;
        height: 56px;
        padding: 0px 0px 0.1em;
    }
`;

const Item = posed.li({
    open: { x: 0, opacity: 1 },
    closed: { x: -100, opacity: 0 }
});
export const anchor = ({ color }: { color: string }) => css`
    a {
        cursor: pointer;
        position: relative;
        font-weight: bold;
        color: ${({ theme: { colors } }) => colors.black};
        text-transform: uppercase;
        font-size: 1.5em;
        text-decoration: none;
        transition: opacity 0.3s ease 0s;
        border-bottom: 1px solid transparent;
        @media ${devices.desktop} {
            color: ${color};
            font-size: 0.85em;
            letter-spacing: 0.8px;
            height: 56px;
            padding: 0px 0px 0.1em;
        }
        &.active {
            text-decoration: underline;
        }
    }
`;
export const MenuItem = styled(Item)<{ isFixed: boolean }>`
    opacity: 0;
    color: white;
    letter-spacing: 0.8px;
    height: 56px;
    padding: 0px 0px 0.1em;
    ${({ isFixed, theme: { colors } }) =>
        anchor({ color: isFixed ? colors.black : colors.white })}
    @media ${devices.desktop} {
        display: inline-block;
        transform: none !important;
        opacity: 1 !important;
        transition: none 0s ease 0s;
        margin: 0px 0px 0px 1.25em;
        
    }
`;
