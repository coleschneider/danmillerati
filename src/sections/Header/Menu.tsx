import React from "react";
import styled from "styled-components";
import devices from "../../theme/devices";

const MenuBar = styled.span`
    position: absolute;
    display: block;
    width: 40%;
    height: 2px;
    left: 30%;
    transition: all 0.3s ease 0s;
    background: ${({ theme: { colors } }) => colors.black};
`;
const BarOne = styled(MenuBar)`
    top: ${({ isOpen }: { isOpen: boolean }) => (isOpen ? "29px" : "23px")};
    transform: ${({ isOpen }: { isOpen: boolean }) =>
        isOpen ? "rotate(45deg)" : "rotate(0)"};
`;
const BarTwo = styled(MenuBar)`
    transform: ${({ isOpen }: { isOpen: boolean }) =>
        isOpen ? "23px" : "25px"};
    transform: ${({ isOpen }: { isOpen: boolean }) =>
        isOpen ? "rotate(-45deg)" : "rotate(0)"};
`;
const MenuToggle = styled.div`
    position: absolute;
    z-index: 999;
    top: 0px;
    right: 0px;
    padding: 0px;
    margin: 0px;
    transition: transform 0.3s ease 0s;
    @media ${devices.desktop} {
        display: none;
    }
    @media ${devices.desktopMin} {
        right: 1em;
    }
`;
const MenuIcon = styled.button`
    cursor: pointer;
    position: relative;
    float: right;
    width: 3.5rem;
    outline: none;
    height: 3.5rem;
    color: ${({ theme: { colors } }) => colors.black};
    font-family: inherit;
    font-size: inherit;
    -webkit-appearance: none;
    resize: none;
    background: none;
    border-width: initial;
    border-style: none;
    border-color: initial;
    border-image: initial;
    border-radius: 0px;
`;

interface Props {
    isOpen: boolean;
    onClick: () => void;
}
const Menu = ({ isOpen, onClick }: Props) => {
    return (
        <MenuToggle>
            <MenuIcon onClick={onClick}>
                <BarOne isOpen={isOpen} />
                <BarTwo isOpen={isOpen} />
            </MenuIcon>
        </MenuToggle>
    );
};

export default Menu;
