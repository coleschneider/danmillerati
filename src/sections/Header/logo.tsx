import * as React from "react";
import styled from "styled-components";
import devices from "../../theme/devices";
import Icons from "../../theme/Icons";

interface Props extends HeaderProps {
    width: string;
    maxWidth: string;
    fill: string;
    height: string;
    margin: string;
}

const SVG = styled.svg<Partial<Props>>`
    width: ${({ width }) => (!width ? "68px" : width)};
    height: ${({ height }) => height};
    fill: ${({ theme: { colors }, fill }) => (!fill ? colors.black : fill)};
    max-width: ${({ maxWidth }) => maxWidth};
    margin: ${({ margin }) => margin};
    overflow: hidden;
    @media ${devices.desktop} {
        fill: ${({ isFixed, theme: { colors } }) =>
            isFixed ? colors.black : colors.white};
    }
`;

const Svg = ({
    height,
    margin,
    isFixed,
    width,
    maxWidth,
    fill,
    iconType,
    viewBox
}: Partial<Props> & { iconType: IconTypes; viewBox: string }) => (
    <SVG
        width={width}
        height={height}
        maxWidth={maxWidth}
        fill={fill}
        isFixed={isFixed}
        margin={margin}
        xmlns="http://www.w3.org/2000/svg"
        viewBox={viewBox}
    >
        {Icons[iconType]}
    </SVG>
);
export default Svg;
