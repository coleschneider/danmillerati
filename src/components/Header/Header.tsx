import * as React from "react";
import styled from "styled-components";
import posed from "react-pose";
import { withRouter, RouteComponentProps } from "react-router-dom";
import devices from "../../theme/devices";
import Menu from "./Menu";
import Navigation from "./Navigation";
import useScroll from "../../hooks/useScroll";

const HeadroomWrapper = styled.div`
    vertical-align: baseline;
    margin: 0;
    padding: 0;
    border-width: 0;
    border-style: initial;
    border-color: initial;
    border-image: initial;
    font: inherit;
`;
const HeadroomUnfixed = styled.div`
    margin: 0;
    padding: 0;
    border-width: 0;
    border-style: initial;
    border-color: initial;
    border-image: initial;
    font: inherit;
    min-height: 0;
    min-width: 0;
    box-sizing: border-box;
    display: block;
`;
const PosedGroupNav = posed.nav({
    open: {
        height: "auto",
        delayChildren: 200,
        staggerChildren: 50
    },
    closed: {
        height: "56px",
        staggerChildren: 50,
        delay: 300
    }
});
const MenuOverlay = styled.div<Partial<HeaderProps>>`
    position: fixed;
    top: 0;
    left: 0;
    z-index: 5;
    display: block;
    background: ${({ theme: { colors } }) => colors.black};
    transition: all 0.56s cubic-bezier(0.52, 0.16, 0.24, 1) 0s;
    @media ${devices.desktopMin} {
        display: none;
    }
`;
interface FixedNav {
    isFixed: boolean;
    isOpen: boolean;
}

const Nav = styled(PosedGroupNav)<FixedNav>`
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    background: ${({ theme: { colors } }) => colors.white};
    left: 0;
    width: 100%;
    z-index: 5;
    overflow-y: hidden;
    transition: background 0.3s ease 0s, border 0.3s ease 0s;
    border-bottom: ${({ theme: { colors } }) => `1px solid ${colors.grey}`};
    overscroll-behavior: contain contain;
    @media ${devices.desktop} {
        line-height: 56px;
        border: ${({ isFixed, theme: { colors } }) =>
            isFixed ? `1px solid ${colors.grey}` : "1px solid transparent"};
        background: ${({ isFixed, theme: { colors } }) =>
            isFixed ? colors.white : "transparent"};
        max-height: 56px;
        bottom: auto;
    }
`;

const Header = (props: RouteComponentProps) => {
    const [isOpen, toggle] = React.useState(false);
    const isFixed = useScroll("nav-id");

    const toggleMenu = () => toggle(!isOpen);
    return (
        <HeadroomWrapper>
            <HeadroomUnfixed>
                <MenuOverlay isOpen={isOpen} />
                <Nav
                    pose={isOpen ? "open" : "closed"}
                    isOpen={isOpen}
                    isFixed={isFixed}
                    id="nav-id"
                >
                    <Menu isOpen={isOpen} onClick={toggleMenu} />
                    <Navigation
                        {...props}
                        toggle={toggleMenu}
                        isOpen={isOpen}
                        isFixed={isFixed}
                    />
                </Nav>
            </HeadroomUnfixed>
        </HeadroomWrapper>
    );
};
export default withRouter(Header);
