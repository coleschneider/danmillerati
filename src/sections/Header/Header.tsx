import * as React from "react";
import styled from "styled-components";
import devices from "../../theme/devices";
import useScroll from "../../hooks/useScroll";
// eslint-disable-next-line
import Navigation from "./Navigation";
import Menu from "./Menu";

const HeadroomWrapper = styled.div`
    vertical-align: baseline;
    margin: 0;
    padding: 0;
    border-width: 0;
    border-style: initial;
    border-color: initial;
    border-image: initial;
    font: inherit;
`;
const HeadroomUnfixed = styled.div`
    margin: 0;
    padding: 0;
    border-width: 0;
    border-style: initial;
    border-color: initial;
    border-image: initial;
    font: inherit;
    min-height: 0;
    min-width: 0;
    box-sizing: border-box;
    display: block;
`;

const MenuOverlay = styled.div<Partial<HeaderProps>>`
    height: 100%;
    width: 100%;
    position: fixed;
    top: 0;
    left: 0;
    z-index: 99;
    opacity: ${({ isOpen }) => (isOpen ? "0.8" : "0")};
    visibility: ${({ isOpen }) => (isOpen ? "visible" : "hidden")};
    display: block;
    background: ${({ theme: { colors } }) => colors.black};
    transition: all 0.56s cubic-bezier(0.52, 0.16, 0.24, 1) 0s;

    @media ${devices.desktop} {
        display: none;
    }
`;
interface FixedNav {
    isFixed: boolean;
    isOpen: boolean;
}

const Nav = styled.nav<FixedNav>`
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    max-height: ${({ isOpen }) => (isOpen ? "100%" : "56px")};
    width: 100%;
    z-index: 99;
    overflow-y: hidden;
    transition: max-height 0.56s cubic-bezier(0.52, 0.16, 0.24, 1) 0s,
        background 0.3s ease 0s, border 0.3s ease 0s;
    border-bottom: ${({ theme: { colors } }) => `1px solid ${colors.grey}`};
    overscroll-behavior: contain contain;

    @media ${devices.desktop} {
        border: ${({ isFixed, theme: { colors } }) =>
            isFixed ? `1px solid ${colors.grey}` : "1px solid transparent"};

        background: ${({ isFixed, theme: { colors } }) =>
            isFixed ? colors.white : "transparent"};

        /* max-height: none/s; */
        bottom: auto;
        line-height: 56px;
    }
`;

const Header = () => {
    const [isOpen, toggle] = React.useState(false);
    const isFixed = useScroll("nav-id");

    const toggleMenu = () => toggle(!isOpen);
    return (
        <HeadroomWrapper>
            <HeadroomUnfixed>
                <MenuOverlay isOpen={isOpen} />
                <Nav isFixed={isFixed} isOpen={isOpen} id="nav-id">
                    <Menu isOpen={isOpen} onClick={toggleMenu} />
                    <Navigation isOpen={isOpen} isFixed={isFixed} />
                </Nav>
            </HeadroomUnfixed>
        </HeadroomWrapper>
    );
};
export default Header;
