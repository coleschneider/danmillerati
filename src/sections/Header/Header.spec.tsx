import * as React from "react";
import Header from "./Header";

describe("Header", () => {
    let wrapper;
    const windowSpy = jest.fn();
    beforeAll(() => {
        window.addEventListener("resize", windowSpy);
    });

    it("should match the snapshot", () => {
        // @ts-ignore
        window.innerWidth = 320;
        window.dispatchEvent(new Event("resize"));
        wrapper = shallow(<Header />);
        console.log(wrapper.debug(), windowSpy, window.innerWidth);
    });
});
