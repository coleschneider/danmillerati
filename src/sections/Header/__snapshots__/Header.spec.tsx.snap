// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot 1`] = `
<Header__HeadroomWrapper>
  <Header__HeadroomUnfixed>
    <Header__MenuOverlay
      isOpen={false}
    />
    <Header__Nav
      id="nav-id"
      isFixed={false}
      isOpen={false}
    >
      <Menu
        isOpen={false}
        onClick={[Function]}
      />
      <Navigation
        isFixed={false}
        isOpen={false}
        toggle={[Function]}
      />
    </Header__Nav>
  </Header__HeadroomUnfixed>
</Header__HeadroomWrapper>
`;
